### ISSUE TITLE
File Analysis Too Slow

### ISSUE THREAD

**Issue by y1cunhui** [microsoft/pyright#5475]
# Files
My repo only include 2 python files: `evm_codes.py` and `evm.py`, it's a simple stack-based virtual machine simulator.
`evm_codes.py` is like:
```
STOP = 0x0
ADD = 0x01
MUL = 0x02
SUB = 0x03
DIV = 0x04
SDIV = 0x05
MOD = 0x06
SMOD = 0x07
ADDMOD = 0x08
MULMOD = 0x09
EXP = 0x0a
SIGNEXTEND = 0x0b
LT = 0x10
......
```
Just the declarations of Opcodes.

`evm.py` is like:
```
import json
import os
import evm_codes

# some simple helper functions

def evm(code):
    pc = 0
    success = True
    stack = []
    memory = []
    
    while pc < len(code):
        op = code[pc]
        pc += 1

        match op:
            case evm_codes.STOP:
                break
            case evm_codes.ADD:
                stack = [stack[0] + stack[1]] + stack[2:]
                if stack[0] > (2 ** 256) - 1:
                    stack[0] -= (2 ** 256)
            case evm_codes.MUL:
                stack = [stack[0] * stack[1]] + stack[2:]
                if stack[0] > (2 ** 256) - 1:
                    stack[0] %= (2 ** 256)
            case evm_codes.SUB:
                stack = [stack[0] - stack[1]] + stack[2:]
                if stack[0] < 0:
                    stack[0] += (2 ** 256)
            case evm_codes.DIV:
                if stack[1] == 0:
                    stack = [0] + stack[2:]
                else:
                    stack = [stack[0] // stack[1]] + stack[2:]
            ......
```
The main function is just a loop and match-case, and each case only include simple operations.

# Question
When I just add an opcode in `evm_codes.py`, it takes 1 min to analyze the 2 files. And in this process, it always give the wrong code autocomplete which affect me a lot.

# Logs:
```
2023-07-11 11:29:56.393 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:29:56.394 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:29:57.238 [info] (1606) Background analysis message: setFileOpened
2023-07-11 11:29:57.238 [info] (1606) Background analysis message: markFilesDirty
2023-07-11 11:29:57.461 [info] (1606) Background analysis message: setFileOpened
2023-07-11 11:29:57.462 [info] (1606) Background analysis message: markFilesDirty
2023-07-11 11:29:57.682 [info] (1606) Background analysis message: getSemanticTokens delta
2023-07-11 11:29:57.689 [info] (1606) [BG(1)] getSemanticTokens delta previousResultId:1689046121853 at /home/y1cunhui/evm-from-scratch/python/evm_codes.py ...
2023-07-11 11:29:57.689 [info] (1606) [BG(1)]   parsing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (7ms)
2023-07-11 11:29:57.692 [info] (1606) [BG(1)]   binding: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (2ms)
2023-07-11 11:29:57.698 [info] (1606) [BG(1)] getSemanticTokens delta previousResultId:1689046121853 at /home/y1cunhui/evm-from-scratch/python/evm_codes.py (15ms)
2023-07-11 11:29:57.705 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:29:57.711 [info] (1606) [FG] parsing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (5ms)
2023-07-11 11:29:57.712 [info] (1606) [FG] binding: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (1ms)
2023-07-11 11:29:57.713 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:29:57.956 [info] (1606) Background analysis message: analyze
2023-07-11 11:29:58.773 [info] [Info  - 11:29:58 AM] (1606) SourceFile: Received fs event 'change' for path '/home/y1cunhui/evm-from-scratch/python/evm_codes.py'
2023-07-11 11:29:58.779 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm.py ...
2023-07-11 11:29:58.779 [info] (1606) [BG(1)]   checking: /home/y1cunhui/evm-from-scratch/python/evm.py (822ms)
2023-07-11 11:29:58.779 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm.py (823ms)
2023-07-11 11:29:58.780 [info] (1606) [BG(1)] indexing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py [found 101] (1ms)
2023-07-11 11:29:58.780 [info] (1606) Indexing Done: /home/y1cunhui/evm-from-scratch/python/evm_codes.py
2023-07-11 11:29:58.781 [info] (1606) Background analysis message: markFilesDirty
2023-07-11 11:29:58.794 [info] (1606) Background analysis message: analyze
2023-07-11 11:30:50.544 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm.py ...
2023-07-11 11:30:50.545 [info] (1606) [BG(1)]   checking: /home/y1cunhui/evm-from-scratch/python/evm.py (51750ms)
2023-07-11 11:30:50.545 [info] [Info  - 11:30:50 AM] (1606) [BG(1)] Long operation: checking: /home/y1cunhui/evm-from-scratch/python/evm.py (51750ms)
2023-07-11 11:30:50.545 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm.py (51750ms)
2023-07-11 11:30:50.545 [info] [Info  - 11:30:50 AM] (1606) [BG(1)] Long operation: analyzing: /home/y1cunhui/evm-from-scratch/python/evm.py (51750ms)
2023-07-11 11:30:50.546 [info] (1606) Background analysis message: resumeAnalysis
2023-07-11 11:30:50.550 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py ...
2023-07-11 11:30:50.550 [info] (1606) [BG(1)]   checking: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (4ms)
2023-07-11 11:30:50.550 [info] (1606) [BG(1)] analyzing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py (4ms)
2023-07-11 11:30:50.551 [info] (1606) Background analysis message: resumeAnalysis
2023-07-11 11:30:50.553 [info] (1606) [BG(1)] indexing: /home/y1cunhui/evm-from-scratch/python/evm_codes.py [found 101] (1ms)
2023-07-11 11:30:50.553 [info] (1606) Indexing Done: /home/y1cunhui/evm-from-scratch/python/evm_codes.py
2023-07-11 11:30:50.823 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:30:50.824 [info] (1606) Background analysis message: getDiagnosticsForRange
2023-07-11 11:33:17.267 [info] (1606) Background analysis message: getDiagnosticsForRange
```

It takes 51s to analyze the `evm.py`. However, I only import `json` and `os` packages, and this file only include 200+ lines of simple code, so I'm confused why it takes so long to analyze this file...

It takes long time on my PC and laptop, so maybe I think it's not bacause of the bad performance of my computer.

**Comment by erictraut** [comment]
@StellaHuang95, I expanded your access to the pyright repo to "maintainer", so you should now be able to transfer to/from pyright.

**Comment by erictraut** [comment]
Thanks for providing a link to your code. 

This code presents a pathological case for pyright. The function is not complex enough (in terms of cycolmatic complexity) to meet the "too complex to analyze" threshold, but it's just complex enough to cause problems. The biggest issue is with the `stack` variable which is unannotated and is assigned 80 values in the function. Most of these values depend on `stack` for type evaluation, and the loop guarantees that all `stack` usages are dependent on all others for type convergence. Many of the expressions also include subexpressions of the form `stack[n]` where `n` is an int literal. These subexpressions are candidates for type narrowing in pyright's code flow engine, which exacerbates the performance issue.

If I add a `list[Any]` annotation to the `stack` variable, the analysis time drops from 35s to 6s on my machine. If I add an `Any` annotation to the `stack` variable, the analysis time drops to less than 1s.

I've attempted to add some heuristics to catch this class of pathological behavior and handle it in a more optimal manner. This results in <1s analysis times even if a type annotation is omitted for `stack`.

**Comment by y1cunhui** [comment]
Really thanks for your effort and the detailed explanation! This also gives me more understanding of the type annotation/evaluation system.

**Comment by erictraut** [comment]
We leave issues open until they are included in the next published release. Reopening for now.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. It will also be included in a future release of pylance.
